lib/msf/core/auxiliary/masscan.rb

```diff
@@ -0,0 +1,277 @@
+  ], Auxiliary::Nmap)
+
+  deregister_options("RPORT")
+  @masscan_args = []
+  @masscan_bin = masscan_binary_path
+end
+
+def rports
+  datastore['RPORTS']
+end
+
+def rport
+  datastore['RPORT']
+end
+
+def set_masscan_cmd
+  self.masscan_bin || (raise RuntimeError, "Cannot locate massscan binary")
+  masscan_set_log
+  masscan_add_ports
+  masscan_cmd = [self.nmap_bin]
+  self.masscan_args.unshift("-oX #{self.masscan_log[1]}")
+  masscan_cmd << self.masscan_args.join(" ")
+  masscan_cmd << datastore['RHOSTS']
+  masscan_cmd.join(" ")
+end
+
+def get_masscan_ver
+  self.masscan_bin || (raise RuntimeError, "Cannot locate nmap binary")
+  res = ""
+  masscan_cmd = [self.masscan_bin]
+  masscan_cmd << "--version"
+  res << %x{#{masscan_cmd.join(" ")}} rescue nil
+  res.gsub(/[\x0d\x0a]/n,"")
+end
+
+# Takes a version string in the form of Major.Minor and compares to
+# the found version. It yells at you specifically if you try to
+# compare a float b/c that's going to be a super common error.
+# Comparing an Integer is okay, though.
+def masscan_version_at_least?(test_ver=nil)
+  raise ArgumentError, "Cannot compare a Float, use a String or Integer" if test_ver.kind_of? Float
+  unless test_ver.to_s[/^([0-9]+(\x2e[0-9]+)?)/n]
+    raise ArgumentError, "Bad Nmap comparison version: #{test_ver.inspect}"
+  end
+  test_ver_str = test_ver.to_s
+  tnum_arr = $1.split(/\x2e/n)[0,2].map {|x| x.to_i}
+  installed_ver = get_nmap_ver()
+  vtag = installed_ver.split[2] # Should be ["Nmap", "version", "X.YZTAG", "(", "http..", ")"]
+  return false if (vtag.nil? || vtag.empty?)
+  return false unless (vtag =~ /^([0-9]+\x2e[0-9]+)/n) # Drop the tag.
+  inum_arr = $1.split(/\x2e/n)[0,2].map {|x| x.to_i}
+  return true if inum_arr[0] > tnum_arr[0]
+  return false if inum_arr[0] < tnum_arr[0]
+  inum_arr[1].to_i >= tnum_arr[1].to_i
+end
+
+def masscan_build_args
+  raise RuntimeError, "nmap_build_args() not defined by #{self.refname}"
+end
+
+def masscan_run
+  masscan_cmd = set_masscan_cmd
+  begin
+    masscan_pipe = ::Open3::popen3(masscan_cmd)
+    @masscan_pid = masscan_pipe.last.pid
+    print_status "Masscan: Starting masscan with pid #{@masscan_pid}"
+    temp_masscan_threads = []
+    temp_masscan_threads << framework.threads.spawn("Module(#{self.refname})-MasscanStdout", false, masscan_pipe[1]) do |np_1|
+      np_1.each_line do |masscan_out|
+        next if masscan_out.strip.empty?
+        print_status "Masscan: #{masscan_out.strip}" if datastore['MASSCAN_VERBOSE']
+      end
+    end
+
+    temp_masscan_threads << framework.threads.spawn("Module(#{self.refname})-MasscanStderr", false, masscan_pipe[2]) do |np_2|
+      np_2.each_line do |masscan_err|
+        next if masscan_err.strip.empty?
+        print_status  "Massscan: '#{masscan_err.strip}'"
+      end
+    end
+
+    temp_masscan_threads.map {|t| t.join rescue nil}
+    masscan_pipe.each {|p| p.close rescue nil}
```

Comment:
This should be done in an `ensure`, otherwise these will not close if something between your `popen3` call and now throws an unhandled exception.